### DISCUSSION TITLE
Genbank database

### DISCUSSION THREAD

**Discussion by marchoeppner** [bio-raum/FooDMe2#26] [General]
Just a general comment on Genbank:

The latest version of the eukaryote branch of nt is a whooping 650GB when fully downloaded. Unsurprisingly, running it consumes a pretty substantial amount of RAM (60GB+ on 8 cores). It seems to me that this is a something that will create issues for many users. I have already documented that Genbank is optional and can be skipped during installation. 

Something we should keep an eye on in any case -  and a good thing we have these single gene databases available.

**Comment by marchoeppner** [comment]
Addendum - I don't think we can even run this database ;) It blows out 256GB of Ram. The issue is, I think, that the database size is in fact larger than the available memory on the machine (650Gb vs 256Gb). With 50 ASVs from Dada, it just keeps eating up RAM without finishing (at least it didn't finish within 30mins, which in itself is already unacceptable). 

Seems to me that using Genbank isn't going to be an option moving forward. What do you reckon?

**Reply by marchoeppner** [reply]
Fun fact, Genbank today is 8 times the size of what it was just four years ago^^

**Reply by marchoeppner** [reply]
Another observation - I can run Foodme1 against the new version of Genbank NT. It also consumes a ton of Ram (>>100GB), but it still manages to get the results before the node starts swapping. Not sure why that is - may need some additional investigation...

**Reply by marchoeppner** [reply]
Right, so the reason was a bit "stupid"... we set "eukaryota" as the taxonomic root if no other taxid is specified as filter. For the pre-configured primer set, the taxid_filter is then switched to "amniota" - all good. But the primer set also specifies Midori lRNA as the database. Running without a pre-configured primer set to test Genbank hence not only requires the primer sequences as fasta, but should also go hand in hand with a user-specified taxid_filter. Which is not requested nor otherwise evaluated - and results in Blast jobs that query all of Genbank nt -> RAM goes "boom". 

With amniotes set as taxid_filter from the command line , the Blast job "only" requires 80GB of Ram. 

Will add a warning.

**Comment by gregdenay** [comment]
Hi Marc, the colleagues in Münster made me aware that there is a new BLAST db. I think NCBI noticed that NT is grwing out of control, the new db is core_nt and can be downloaded from the NCBI FTP, it's 'only' about 150Gb.

That would probably be a good alternative?

**Reply by gregdenay** [reply]
Moe info here
https://ncbiinsights.ncbi.nlm.nih.gov/2024/07/18/new-blast-core-nucleotide-database/

**Reply by marchoeppner** [reply]
Seems worth a try at least. 600GB is a lot to ask from users, let alone the RAM requirements. Will download later today and see if that new database works.

**Reply by marchoeppner** [reply]
After reading the explanation, I think it makes sense to simply replace the current "genbank euk" with "core_nt". It is a lot smaller (still > 200GB...) and will be the NCBI default moving forward, so why not. 

Updated the resources and the genbank installation script. My local test runs through no problem on this. I kept the db label "genbank" tho.

**Comment by gregdenay** [comment]
Genbank installed succesfully after your fix. I'll test it today.
The work folder keeps a copy of everything ( about 250Gb) when running build_references with the genbank db.
I tried `nextflow clean -f` to remove this but it doesn't seem to do anything, one actually needs to manually delete the work folder.
Is there an option to run a small script after pipeline successful completion to just delete the unescessary archives? Something like snakemake `onsuccess` (https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#onstart-onsuccess-and-onerror-handlers)

**Reply by marchoeppner** [reply]
I think the issue here is that "cleanup" only works on the outputs of individual processes. The stuff that gets left behind in this case are the staged files from Genbank etc. Basically, when we do `file(path_to_refseq_mito)` it silently pulls that file (no need to use wget) and stores it under work/staged-files. If we were to write a bunch of wget processes, the `clean` command would presumably "work" as expected. Not sure that is a very "nice" solution tho. 

Looks like someone already suggested a fix for this (https://github.com/nextflow-io/nextflow/issues/2139) but with no reaction from the developers. 

So at the moment, I think, the only viable solution is the one you chose, i.e. deleting the work directory.

**Reply by marchoeppner** [reply]
See: https://github.com/nextflow-io/nextflow/issues/5091

I think we can implement that in some form.

**Reply by marchoeppner** [reply]
Fixed in commit 8e76759
